### ISSUE TITLE
Log graph mis-alignment when abbreviated commit hashes conflict

### ISSUE THREAD

**Issue by phil-s** [magit/magit#3034]
Magit 2.10.1-dirty, Git 2.3.5, Emacs 25.2.1, gnu/linux

In this example d7e3227 would have been an ambiguous commit hash, and so git has displayed it as d7e3227c instead. This extra character is then breaking the alignment of the graph.

Issue #1425 was a catch-all for issues with log graphs, but has been closed in favour of issue #2989.

This problem is an edge case, so could likewise be closed in favour of #2989, but I have raised it in case you can see a straightforward fix.

What I currently see:

    d7e3227c *
             |\
    7992537 | *
    64d126a * |
             |\ \
    50f307e | * |
    9cfdea0 * | |
             |\ \ \
    5f25729 | * | |
             | | |/
             | |/|
    e88f451 * | |
             |\ \ \
    bb4aa67 | * | |
    69befe6 * | | |
             |\ \ \ \
    e217441 | * | | |
             | | |/ /
             | |/| |
    6f3923f * | | |
             |/ / /
    e8f09c4 * | |
             |\ \ \

How it ought to look:

    d7e3227 *
            |\
    7992537 | *
    64d126a * |
            |\ \
    50f307e | * |
    9cfdea0 * | |
            |\ \ \
    5f25729 | * | |
            | | |/
            | |/|
    e88f451 * | |
            |\ \ \
    bb4aa67 | * | |
    69befe6 * | | |
            |\ \ \ \
    e217441 | * | | |
            | | |/ /
            | |/| |
    6f3923f * | | |
            |/ / /
    e8f09c4 * | |
            |\ \ \

**Comment by kyleam** [comment]
Duplicate of #2351

**Comment by phil-s** [comment]
Additional: The issue of the misalignment occurring *throughout* the log goes away in either of the following scenarios:

* git's `core.abbrev` value is set (to *any* valid number).
* When the HEAD commit of the log is *not* that of the currently checked-out branch.

In these cases, only the specific lines with longer hashes are out of alignment.

**Comment by tarsius** [comment]
d7fcc92ed3af8e817c5521c6f2785aee0a871cd7 explains pretty well what's going on. Basically you have run into "but these situations should be rare and harmless" case. Harmless maybe, but ugly.

Maybe we should determine the abbreviation length for two commits and use the shorter. Also not perfect, but would make this even less likely to happen.

**Comment by phil-s** [comment]
That sounds like a pragmatic workaround.

**Comment by phil-s** [comment]
Would it be wrong to simply post-process the buffer with the standard `align.el` functionality? That could then sort out *all* misalignments (and with no additional calls to git).

**Comment by kyleam** [comment]
I don't think we should spend much time trying to fix this.  If the
user sees this issue, it's an (ugly but harmless) sign that they
should increase the value of `core.abbrev`.  With newer versions of
Git that auto-size the default abbrev length, users will be less
likely to run into this because Git will choose a reasonable size for
them.

**Comment by phil-s** [comment]
Hmm. I kinda disagree with your reasoning (as clashes can absolutely still happen with a "reasonable" size, whether it's set by the user or by git), so I don't think it's a sign that users should increase the value *unless* it's happening regularly.

It's certainly not a particularly important bug, though -- it's purely cosmetic, and a given user might well never encounter it.

**Comment by kyleam** [comment]
> Unless it's happening *regularly* I would disagree [...]

Yeah, I don't disagree with that---but then it's a one-off ugly log,
and, after that commit isn't HEAD, a one-off ugly line of the log.  And
I agree that this is something to keep in mind, especially wrt #2989.
A rewrite like that would also allow us to recondiser where to put the hash.

**Comment by kyleam** [comment]
> Magit 2.10.1-dirty, Git 2.3.5, Emacs 25.2.1, gnu/linux

Just out of curiousity, how does Git >=2.11 do with your example above?

**Comment by phil-s** [comment]
I've just compiled 2.12.2, and I'm a little surprised, actually -- it's chosen a value of 10 for this repo. That does seem pretty safe/conservative. `d7e3227c6c`

**Comment by kyleam** [comment]
Thanks for taking the time to check.  If you want more information on
their scaling approach, [here's the message] that eventually led to
the feature.

[here's the message]: https://public-inbox.org/git/CA+55aFwbCNiF0nDppZ5SuRcZwc9kNvKYzgyd_bR8Ut8XRW_p4Q@mail.gmail.com/

**Comment by tarsius** [comment]
> I don't think we should spend much time trying to fix this.

Fixing it doesn't take long, talking about it does. Should I just merge #3037 and be done with it?

**Comment by kyleam** [comment]
> Should I just merge #3037 and be done with it?

For the reasons I gave [above], I don't think preventing the
mis-alignment is worth an extra call to git.  Though in practice, I
doubt the extra call matters much because we call
`magit-abbrev-length` only once when preparing log buffers.  So,
despite my "wontfix" leaning in this case, merging #3037 is OK with
me.

[above]: https://github.com/magit/magit/issues/3034#issuecomment-289472262
